com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java

```diff
@@ -885,7 +893,8 @@ private static Rig.Model loadModel(Node node, Model model, ArrayList<Modelimport
         }
 
         modelBuilder.setId(MurmurHash.hash64(node.name)); // the node name is the human readable name (e.g Sword)
-        modelBuilder.setLocal(toDDFTransform(node.local));
+        // need to flatten all meshes of the model
+        modelBuilder.setLocal(node.parent != null ? toDDFMulTransforms(node.parent.world, node.local) : toDDFTransform(node.local));
```

Comment:
> First, I don't like the idea of reusing the name "local" for something that clearly isn't local.
> 
> Also, has this been tested at runtime? Or have you only tested it in the editor? E.g. at runtime we use this local matrix together with the bone world matrix.

Two local keywords:

* `setLocal`:
  * I think the name `setLocal` is OK for the `Model`, not for `MeshSet`. Here, the transform is used in both  `Model` and `MeshSet`. The name `setLocal` comes from [`rig_ddf.proto`](https://github.com/defold/defold/blob/4ef3d1b9a792ab83b873347fdb879540f20ae2fa/engine/rig/proto/rig/rig_ddf.proto#L296).
* `node.local`:
  * I don't think it is the wrong name for `node`.
* Yes, I tested it at runtime and editor.
  * Here you can see the screenshot at runtime: https://github.com/defold/defold/issues/9603#issuecomment-2422925085

I found I can change the line:
```java
modelBuilder.setLocal(node.parent != null ? toDDFMulTransforms(node.parent.world, node.local) : toDDFTransform(node.local));
```
To:
```java
modelBuilder.setLocal(node.world);
```
It will be simpler, but the name is a bit strange.

Reply:
> Yes, I tested it at runtime

However, I don't see an animated+rigged character using this? Was that tested?

> setLocal

I wasn't objecting to the name "local", I was objecting to the fact that you are modifying the local matrix to be _not_ local.
By applying the parents world transform to it, you are making it a _world_ transform, but store it in a variable named "local". That's not good.

Reply:
> However, I don't see an animated+rigged character using this? Was that tested?

I tested some animated+rigged models in my project. It's no problem.
You are right. I have to build another model that contains animated+rigged and child nodes.

> By applying the parents world transform to it, you are making it a world transform, but store it in a variable named "local". That's not good.

I know what you mean. The `loadModelInstances` is an iteration function and it flattens the node tree to a model list. So I have passed the node's world as the model's local. Maybe we need to add a `world` parameter for `Model` in the `rig_ddf.proto`. Or keep the node tree by making the `Model` to be a tree structure.

Reply:
I test the animated+rigged+child model. It is no problem.

[gltf-animated+rigged+child.zip](https://github.com/user-attachments/files/17478479/gltf-animated%2Brigged%2Bchild.zip)

> Blender file is **gltf-animated+rigged+child.zip/resources/sample.blend**

![gltf-animated+rigged+child](https://github.com/user-attachments/assets/b6ad5a56-eb94-43a7-982a-f1b8f7048861)

You can see the `Torus` is a child of `Cube`. The `Cube` is bound with the `head` bone. The `Torus` is bound with the `body` bone. A part of the `Sphere` is bound with the `body` bone and the rest is bound with the `tail` bone.

Reply:
Display in the editor:

![image](https://github.com/user-attachments/assets/36c215a2-0bf1-4335-bbf7-794005812ab1)

Reply:
Here is my thought:

According the [`rig_ddf.proto`](https://github.com/defold/defold/blob/dev/engine/rig/proto/rig/rig_ddf.proto) file:

|dmRigDDF||glTF||
|-|-|-|-|
|[`Model`](https://github.com/defold/defold/blob/43e73233fb7ee712e5e7f40149b2b79ad0f2fcbe/engine/rig/proto/rig/rig_ddf.proto#L292) contains a `Mesh` list|<=>|[Scene](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#reference-scene) contains a `Node` list|Same|
|[`Mesh`](https://github.com/defold/defold/blob/43e73233fb7ee712e5e7f40149b2b79ad0f2fcbe/engine/rig/proto/rig/rig_ddf.proto#L266) no children|**?**=>|[Node](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#reference-node) contains a children|Different|

## Question

How do I convert glTF's nodes(tree) to `dmRigDDF::Model::meshes`(list)?

### Solution 1

Like my PR, set the `meshes` of `Model` to Flatten all nodes of the scene using the world transform.

### Solution 2

Add the `local` parameter for `Mesh`.

Reply:
I try to add a `children` parameter for `Model`. The editor raises a stackoverflow exception when opening the glb file from the editor.

Below is my change.

```proto
message Model // E.g. the Node in the Scene
{
    option (struct_align) = true;

    required dmMath.Transform   local       = 1 [(field_align)=true];
    required uint64             id          = 2;    // E.g. "torso", "head". Or simply "character"
    repeated Mesh               meshes      = 3;
    // If set, then this model should be transformed as a child of the bone
    optional uint64             bone_id     = 4 [default = 0]; // hash of bone id
    repeated Model              children    = 5;
}
```

The reason is called iteratively `pb-value->clj-fn` when scanning the `children`(a `Model` list) of `Model`.

https://github.com/defold/defold/blob/f92f6dd2b4383701190badd9de7804ee90e33c4c/editor/src/clj/editor/protobuf.clj#L499

I am a newbie to Clojure.
How can I fix it?

Reply:
> Maybe we need to add a world parameter for Model in the rig_ddf.proto

This is where my mind is as well. I think this would keep the meshes "flattened", but still have the possibility to use the world transforms when necessary.

Reply:
OK. Let me try.

Reply:
I add a world transform into `Model`.

https://github.com/defold/defold/blob/5cf34975ca9c71fc8d1b9c96b1c1498f87225007/engine/rig/proto/rig/rig_ddf.proto#L292

```protobuf
message Model // E.g. the Node in the Scene
{
    option (struct_align) = true;

    required dmMath.Transform   local       = 1 [(field_align)=true];
    required uint64             id          = 2;    // E.g. "torso", "head". Or simply "character"
    repeated Mesh               meshes      = 3;
    // If set, then this model should be transformed as a child of the bone
    optional uint64             bone_id     = 4 [default = 0]; // hash of bone id
    required dmMath.Transform   world       = 5 [(field_align)=true];
}
```

Then change two lines below: `m_Local` to `m_World`

https://github.com/defold/defold/blob/9679c4c1fe794f86dea317b1da6af959b1577f9c/engine/gamesys/src/gamesys/components/comp_model.cpp#L1413

https://github.com/defold/defold/blob/9679c4c1fe794f86dea317b1da6af959b1577f9c/engine/gamesys/src/gamesys/components/comp_model.cpp#L1417

```cpp
                dmVMath::Matrix4 model_matrix;
                if (render_item->m_BoneIndex != dmRig::INVALID_BONE_INDEX)
                {
                    dmRig::BonePose bone_pose = pose[render_item->m_BoneIndex];
                    model_matrix = dmTransform::ToMatrix4(bone_pose.m_World) * dmTransform::ToMatrix4(render_item->m_Model->m_World); //< `m_Local` to `m_World`
                }
                else
                {
                    model_matrix = dmTransform::ToMatrix4(render_item->m_Model->m_World); //< `m_Local` to `m_World`
                }

                dmVMath::Matrix4 world_matrix     = c->m_World * model_matrix;
                dmVMath::Matrix4 normal_matrix    = dmRender::GetNormalMatrix(render_context, world_matrix);
                bool has_custom_vertex_attributes = vx_decl != world->m_VertexDeclaration;

                uint32_t material_index = render_item->m_MaterialIndex;
                vb_end = WriteWorldSpaceVertexData(world, render_context, c, render_item, &material_infos_vertex, vertex_stride, material_index, world_matrix, normal_matrix, has_custom_vertex_attributes, vb_end);
```

I think it is OK.

Reply:
Alright, I just committed new push - https://github.com/defold/defold/pull/9616/commits/7fb10bc710cc4cefcdbfcdcfd1543dccd050e017